internal/v4/auth_test.go

```diff
@@ -820,6 +874,118 @@ func (s *authSuite) TestDelegatableMacaroonWithBasicAuth(c *gc.C) {
 	})
 }
 
+func (s *authSuite) TestDelegatableMacaroonWithTerms(c *gc.C) {
+	// Create a new server with a third party discharger.
+	s.discharge = dischargeForUser("bob")
+	// User has agreed to terms.
+	s.dischargeTerms = func(s1, s2 string) ([]checkers.Caveat, error) {
+		return nil, nil
+	}
+
+	id1 := newResolvedURL("~charmers/precise/terms-10", 10)
+	err := s.store.AddCharmWithArchive(id1, storetesting.Charms.CharmDir("terms"))
+	c.Assert(err, jc.ErrorIsNil)
+
+	id2 := newResolvedURL("~charmers/precise/terms-11", 11)
+	err = s.store.AddCharmWithArchive(id2, storetesting.Charms.CharmDir("terms"))
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.store.SetPerms(&id1.URL, "read", "bob")
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = s.store.SetPerms(&id2.URL, "read", "bob")
+	c.Assert(err, jc.ErrorIsNil)
+
+	client := httpbakery.NewHTTPClient()
+
+	now := time.Now()
+	var gotBody json.RawMessage
+	httptesting.AssertJSONCall(c, httptesting.JSONCallParams{
+		Handler:  s.srv,
+		URL:      storeURL("delegatable-macaroon"),
+		JSONBody: &id2.URL,
+		ExpectBody: httptesting.BodyAsserter(func(c *gc.C, m json.RawMessage) {
+			gotBody = m
+		}),
+		Do:           bakeryDo(nil),
+		ExpectStatus: http.StatusOK,
+	})
+
+	c.Assert(gotBody, gc.NotNil)
+	var m macaroon.Macaroon
+	err = json.Unmarshal(gotBody, &m)
+	c.Assert(err, gc.IsNil)
+
+	caveats := m.Caveats()
+	foundExpiry := false
+	for _, cav := range caveats {
+		cond, arg, err := checkers.ParseCaveat(cav.Id)
+		c.Assert(err, gc.IsNil)
+		switch cond {
+		case checkers.CondTimeBefore:
+			t, err := time.Parse(time.RFC3339Nano, arg)
+			c.Assert(err, gc.IsNil)
+			c.Assert(t, jc.TimeBetween(now.Add(v4.DelegatableMacaroonExpiry), now.Add(v4.DelegatableMacaroonExpiry+time.Second)))
+			foundExpiry = true
+		}
+	}
+	c.Assert(foundExpiry, jc.IsTrue)
+
+	// Now check that we can use the obtained macaroon to do stuff
+	// as the declared user.
+	s.discharge = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("no discharge")
+	}
+	s.dischargeTerms = func(_, _ string) ([]checkers.Caveat, error) {
+		return nil, errgo.New("no discharge")
+	}
+
+	// First check that we require authorization to access the charm.
+	rec := httptesting.DoRequest(c, httptesting.DoRequestParams{
```

Comment:
I might have missed it, but I'd like to see a test that we can't access the charm with the usual delegatable macaroon (not id-specific) too.